nav2_behavior_tree/plugins/decorator/path_longer_on_approach.cpp

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) 2022 Neobotix GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <string>
+#include <memory>
+#include <vector>
+#include "nav2_util/geometry_utils.hpp"
+
+#include "nav2_behavior_tree/plugins/decorator/path_longer_on_approach.hpp"
+
+namespace nav2_behavior_tree
+{
+
+PathLongerOnApproach::PathLongerOnApproach(
+  const std::string & name,
+  const BT::NodeConfiguration & conf)
+: BT::DecoratorNode(name, conf)
+{
+  node_ = config().blackboard->get<rclcpp::Node::SharedPtr>("node");
+}
+
+bool PathLongerOnApproach::isPathUpdated(
+  nav_msgs::msg::Path & new_path,
+  nav_msgs::msg::Path & old_path)
+{
+  return new_path != old_path && old_path.poses.size() != 0 &&
+         new_path.poses.size() != 0 &&
+         old_path.poses.back() == new_path.poses.back();
+}
+
+bool PathLongerOnApproach::isRobotInGoalProximity(
+  nav_msgs::msg::Path & old_path,
+  double & prox_leng)
+{
+  return nav2_util::geometry_utils::calculate_path_length(old_path, 0) < prox_leng;
+}
+
+bool PathLongerOnApproach::isNewPathLonger(
+  nav_msgs::msg::Path & new_path,
+  nav_msgs::msg::Path & old_path,
+  double & length_factor)
+{
+  return nav2_util::geometry_utils::calculate_path_length(new_path, 0) >
+         length_factor * nav2_util::geometry_utils::calculate_path_length(
+    old_path, 0);
+}
+
+inline BT::NodeStatus PathLongerOnApproach::tick()
+{
+  getInput("path", new_path_);
+  getInput("prox_len", prox_len_);
+  getInput("length_factor", length_factor_);
+
```

Comment:
I think you need to set the state of this node to running when running, and then you can check `if (IDLE) {first_time_ = true}` to reset between execution runs. e.g.

```
if (IDLE) {first_time_ = true}
setStatus(RUNNING)
```

Also `first_time_` defaults to `false` in your hpp file, shouldn't that be `true` and then it gets set to `false` in L86?

Reply:
> I think you need to set the state of this node to running when running, and then you can check if (IDLE) {first_time_ = true} to reset between execution runs.

I'm just not sure about the ideal scenario. But on a hindsight, we can save a feature request in the issue tracker! 

> Also first_time_ defaults to false in your hpp file, shouldn't that be true and then it gets set to false in L86?

Yepp! Solves the above conversation as well.